Save context in between tasks and fix relations migration

moekify opened the pull request:
This PR includes the following changes:

- Fixes from #32 to fix relations mapping
- Saving the context in a separate file for each major tasks that modifies the context
- Fixing m2m interfaces being migrated wrongly to alias fields and not the correct type & interface as v9 should have it
- Sorting collections before migrating data based on if they are junction collections or contain a 'many-to-one' (due to foreign key constraints, we otherwise have likely failures)
- New command line option to pass in context files
- Change 'directus_files_id' in junction table to type 'uuid'

Should fix #35
Should fix #34
Should fix #31 

Includes work from @dwene @hakan-erdem-temiz <3

Next steps: 

- Write a more robust sorting or find an alternative solution

dwene reviewed (approved):
Woo-hoo! I love this, thanks for taking my hacky fixes and making this way more stable!

rijkvanzanten commented:
@moekify Love it! Does this save the file / user ID map to context as well? The data step requires the file/user context map to exist to map the v8 PK to the v9 UUID

moekify commented:
> 
> 
> @moekify Love it! Does this save the file / user ID map to context as well? The data step requires the file/user context map to exist to map the v8 PK to the v9 UUID

Currently, the implementation literally creates a new file for each step completed. Hence, yes, there is a file that contains the user map or file map respectively after the step has completed.

moekify commented:
@rijkvanzanten 
I'd be super happy about a review now. I found the main culprit of mailing migrations to be the foreign key constraints (and the necessity to have items there)

Hence, I implemented some very hacky sort for the collections, so that dependent collections come after others. I believe this will fail with very complex (relations all over the place) schemas. If possible, we should improve this in the future. 

I updated the original description to better reflect all the changes of this PR.

rijkvanzanten commented:
Ahh good one

> Hence, I implemented some very hacky sort for the collections, so that dependent collections come after others. I believe this will fail with very complex (relations all over the place) schemas. If possible, we should improve this in the future.

Alternatively, we could insert all the data first, and then add the relationship / not null constraints last 🤔

rijkvanzanten commented on .gitignore:
```diff
@@ -2,4 +2,15 @@ node_modules
+
+
+# Ignore context files except start
+context/collections.json
+context/data.json
+context/files.json
+context/relations.json
+context/roles.json
+context/schema.json
+context/users.json
+context/completed.json
```
We could put these stateful ones under a folder and ignore the whole folder, like `context/state`

moekify replied:
Done

rijkvanzanten commented on index.js:
```diff
@@ -14,30 +16,75 @@ const commandLineOptions = commandLineArgs([
+  },
 ]);
 
+export async function writeContext(context, section) {
+  context.completedSteps[section] = true;
+	await fs.promises.writeFile(`./context/${section}.json`, JSON.stringify(context));
+}
+
+async function setupContext(context) {
+	const contextJSON = await fs.promises.readFile(
+    commandLineOptions.useContext,
+    'utf8'
+  );
+  console.log('Loading context')
+  const fetchedContext = JSON.parse(contextJSON);
+  Object.entries(fetchedContext).forEach(([key, value]) => {
+    context[key] = value;
+  });
+	console.log(
+		`✨ Loaded context succesfully`
+	)
+}
+
```
lets run a formatter on this (prettier)

moekify replied:
Done!

rijkvanzanten commented on tasks/data.js:
```diff
@@ -29,9 +29,38 @@ async function getCounts(context) {
 	}
 }
 
+// This is definitely a hack to achieve first adding items of collections that have dependencies in other collections i.e m2m, o2m
+// FIXME: Implement a more robust solution to sort collections based on their dependencies, or swap to a different way to seed the data
```
I think we should look into migrating all the data first, and then "finishing" the schema by applying not-null and relationships. That should prevent any foreign key constraint / null constraint issues during the initial migration 🤔

moekify replied:
I actually tried doing this and had the migrating relations step last. 

There is two problems (apparently) with doing so: 

1. The Migrating Data step actually needs to know about the relations 
 - This is actually solvable, as we could easily grab the data about the relations before sending them to the v9 instance
 
2. The data migration fails due to it trying to insert data into a non existent column. (at least my dataset did this) 

More specifically: 

We have a Collection A, which has a m2m relation with Collection B, in a field called CollectionB. 

When I'm now running the data migration step, without having added the relations first, I get an error (from postgres, but probably doesn't matter about the db) that collectionA.collectionB is not a valid column (that can be added data to). I tried to understand what happens here by looking at directus source, but seemingly adding the relations also adjusts how directus (or the db) treats values that arrive there. 

If you can help me solve that riddle, I'd be willing to give this another shot. Otherwise, I already spent quite some time trying this out to no prevail :(

rijkvanzanten replied:
> The data migration fails due to it trying to insert data into a non existent column

Ohhh, because the one-to-many field is virtual and is based on the existence of the row in directus_relations 🤔 Hmm, difficult problem! In order to get around that, we'd almost have to migrate all the data _without_ any of the relational fields, to make sure all the parent items exist, and then apply the relations, and then do another pass over the data to update the items with the relational values 🥴

rijkvanzanten replied:
Or now that I think about it, one-to-many fields shouldn't have to be populated, as they're based on the other m2o value. Maybe if we simply skip over the one to many fields altogether things just work? 🤔

moekify replied:
Oh if that works that would be quite neat. So your suggestion would be to just skip o2m fields and just fill m2o ones? 

I can try this out. 

Probably only going to get around on Sunday to work on this. really busy work week :(

rijkvanzanten replied:
> So your suggestion would be to just skip o2m fields and just fill m2o ones?

Yeah! That should do the trick 🙂 

> Probably only going to get around on Sunday to work on this. really busy work week :(

No worries! Same here 😭

moekify replied:
Hey @rijkvanzanten,
This doesn't really fix the issue when a collection with an m2o is inserted before the corresponding o2m collection, because then it cannot find the right items. 

I can still make this change, but imho it doesn't solve the problem I'm trying to solve with the sorting. 

Wdyt?

rijkvanzanten commented on tasks/relations.js:
```diff
@@ -0,0 +1,72 @@
+          relation.collection_one.startsWith("directus_")) === false
+      );
+    })
+    .map((relation) => ({
+      // @NOTE: one_primary will be removed from Directus soon, so i'm not too worried about it here
+      meta: {
+        many_collection: relation.collection_many,
+        many_field: relation.field_many,
+        many_primary: "id",
+        one_collection: relation.collection_one,
+        one_field: relation.field_one,
+        one_primary: "id",
+        junction_field: relation.junction_field,
+      },
```
many_primary and one_primary have indeed since been removed

moekify replied:
Done

rijkvanzanten commented on tasks/data.js:
```diff
@@ -82,12 +111,12 @@ async function insertBatch(collection, page, context, task) {
-						} else if (systemRelation.one_collection === "directus_files") {
-							item[systemRelation.many_field] =
-								context.fileMap[item[systemRelation.many_field]];
+						if (systemRelation?.meta?.one_collection === "directus_users") {
+							item[systemRelation?.meta?.many_field] =
```
We could rely on `systemRelation.related_collection` / `systemRelation.field` here as well, though it shouldn't matter

rijkvanzanten reviewed (commented):
Looks good to me so far! 

Lets run Prettier on the codebase to get rid of some indentation inconsistencies, and maybe look into the nullable/fk-constraint robustness before merging this. 

Thanks a lot @moekify ❤️

moekify commented:
I'll make the changes @rijkvanzanten by end of the week. Now a bit busy with my normal job haha!

Re changing when to migrate relations, I added a comment to your suggestion. let's discuss there :)

justrealmilk commented:
me and everyone else who is keen as heck to use this but unsure how to contribute to the wizardry all like 👀👀👀

moekify commented:
> 
> 
> me and everyone else who is keen as heck to use this but unsure how to contribute to the wizardry all like 👀👀👀

Feel free to pull my branch and try it out to see if it breaks with the current "sorting" for data migration. Some additional feedback with a different v8 dataset than mine would be actually really useful :)

justrealmilk commented:
> > me and everyone else who is keen as heck to use this but unsure how to contribute to the wizardry all like 👀👀👀
> 
> Feel free to pull my branch and try it out to see if it breaks with the current "sorting" for data migration. Some additional feedback with a different v8 dataset than mine would be actually really useful :)

Gave it a go!

```
TypeError: Cannot read property 'split' of undefined
    at Task.task (file:///root/migrate/migration-tool/tasks/files.js:67:32)
```
https://github.com/moekify/migration-tool/blob/mn/use-context/tasks/files.js#L67

moekify commented:
> 
> 
> > > me and everyone else who is keen as heck to use this but unsure how to contribute to the wizardry all like 👀👀👀
> > 
> > 
> > Feel free to pull my branch and try it out to see if it breaks with the current "sorting" for data migration. Some additional feedback with a different v8 dataset than mine would be actually really useful :)
> 
> Gave it a go!
> 
> ```
> TypeError: Cannot read property 'split' of undefined
>     at Task.task (file:///root/migrate/migration-tool/tasks/files.js:67:32)
> ```
> 
> https://github.com/moekify/migration-tool/blob/mn/use-context/tasks/files.js#L67

Didn't touch the logic of files at all. (also never experienced issues with it) Maybe a different bug or sth with your files is weird :P

justrealmilk commented:
> > > > me and everyone else who is keen as heck to use this but unsure how to contribute to the wizardry all like 👀👀👀
> > > 
> > > 
> > > Feel free to pull my branch and try it out to see if it breaks with the current "sorting" for data migration. Some additional feedback with a different v8 dataset than mine would be actually really useful :)
> > 
> > 
> > Gave it a go!
> > ```
> > TypeError: Cannot read property 'split' of undefined
> >     at Task.task (file:///root/migrate/migration-tool/tasks/files.js:67:32)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > https://github.com/moekify/migration-tool/blob/mn/use-context/tasks/files.js#L67
> 
> Didn't touch the logic of files at all. (also never experienced issues with it) Maybe a different bug or sth with your files is weird :P

Doubt it's my files haha. Decided to clone the repo to my desktop and set everything up so I can console log some stuff but the errors are so intermittent. Sometimes it manages to migrate relations and other times not. The data is static as in it's not being actively modified so I don't know. I'll keep trying

rijkvanzanten commented:
> Decided to clone the repo to my desktop and set everything up

Did you checkout the correct branch too? `mn/use-context`

That relations shape looks like the old one 🙂

justrealmilk commented:
> > Decided to clone the repo to my desktop and set everything up
> 
> Did you checkout the correct branch too? `mn/use-context`
> 
> That relations shape looks like the old one 🙂

Damn it... I did on my VM but when I moved to my desktop I forgot to select a branch!

rijkvanzanten commented:
> Probably only going to get around on Sunday to work on this. really busy work week :(

Also, quick shameless plug here: If y'all would like to work on Directus in a more serious hands-on capacity, consider checking out https://directus.io/careers/ 😄

justrealmilk commented:
> > > > me and everyone else who is keen as heck to use this but unsure how to contribute to the wizardry all like 👀👀👀
> > > 
> > > 
> > > Feel free to pull my branch and try it out to see if it breaks with the current "sorting" for data migration. Some additional feedback with a different v8 dataset than mine would be actually really useful :)
> > 
> > 
> > Gave it a go!
> > ```
> > TypeError: Cannot read property 'split' of undefined
> >     at Task.task (file:///root/migrate/migration-tool/tasks/files.js:67:32)
> > ```
> > 
> > 
> >     
> >       
> >     
> > 
> >       
> >     
> > 
> >     
> >   
> > https://github.com/moekify/migration-tool/blob/mn/use-context/tasks/files.js#L67
> 
> Didn't touch the logic of files at all. (also never experienced issues with it) Maybe a different bug or sth with your files is weird :P

So, I echoed some file objects and noticed none of them had the expected `asset_url` property so I did the next most obvious thing and checked the Directus 8 changelog for API changes. I'm running 8.6.2 and `asset_url` was added in 8.7.0 🤦🏼‍♂️

I've adjusted the code to suit and am testing the v9 API for file imports in Insomnia and receiving
```
{
  "errors": [
    {
      "message": "\"url\" is required",
      "extensions": {
        "code": "INVALID_PAYLOAD"
      }
    }
  ]
}
```

I've looked at the Directus docs and the payload looks fine and the URL works soooo

moekify commented:
@rijkvanzanten Pushed an update, I didn't do anything yet for the o2m/m2m struggle as just omitting the o2m fields wouldn't fix the problem of m2o collections being seeded when the corresponding o2m collection isn't there yet. 

***

@justrealmilk Regarding large files (and many files) and some migration errors there, I also noticed that e.g. Thumbnail generation fails quite easily with this. The reason for that is that directus limits the amount of concurrent thumbnail tasks (can be increased over env variables). With many files coming in concurrently, this clogs up easily and just fails some straight out. Not sure if it's just sufficient to allow more concurrent jobs (will eat RAM/resources).

My suggestion here would be to honestly make an adjustment to the fileupload and just dump all those errors into some kind of errors.json. So you know which files there were issues with and then can either try again for them or resort to something more manual. Imho, this is unrelated to this change though. Feel free to give it a shot :) 

Regarding your payload issues, if this involves a junction collection that relations payload can't be correct. A relation involving a junction collection will always be two different relations (that are relating your tables with the junction table). E.g. from my dataset:

```
{
      "meta": {
        "many_collection": "events_hubs",
        "many_field": "hubs_id",
        "one_collection": "hubs",
        "one_field": "events",
        "junction_field": "events_id"
      },
      "field": "hubs_id",
      "collection": "events_hubs",
      "related_collection": "hubs",
      "schema": null
    },
    {
      "meta": {
        "many_collection": "events_hubs",
        "many_field": "events_id",
        "one_collection": "events",
        "one_field": "hubs",
        "junction_field": "hubs_id"
      },
      "field": "events_id",
      "collection": "events_hubs",
      "related_collection": "events",
      "schema": null
    }
```

rijkvanzanten commented:
> fix the problem of m2o collections being seeded when the corresponding o2m collection isn't there yet.

Does this problem exist? If we're setting up the full schema first, before seeding the actual data, this problem shouldn't happen right?

moekify commented:
> 
> 
> > fix the problem of m2o collections being seeded when the corresponding o2m collection isn't there yet.
> 
> Does this problem exist? If we're setting up the full schema first, before seeding the actual data, this problem shouldn't happen right?

Sorry, my phrasing wasn't very concise. 

I have set up the full schema, including relations etc.

When I now insert data, and I try to insert data into a collection that has a m2o. It will bail on it, unless the corresponding o2m is already filled with data. 

This happens due to it verifying that the item exists that is referenced in the m2o. 

Currently, my solution will somewhat work for this, given I sort m2o to the back. But probably not crazy robust? WDYT? 

Is there anything else, you'd like me to change in here? :)

justrealmilk commented:
So, in the simplest terms, is this a known problem?

Migration tool payload:

```
    url: '/relations',
    method: 'post',
    data: '{"meta":{"many_collection":"portfolio","many_field":"type","many_primary":"id","one_collection":"portfolio_types","one_field":null,"one_primary":"id","junction_field":null},"field":"type","collection":"portfolio","related_collection":"portfolio_types","schema":null}',
```

Directus error: 

`Error: ER_KEY_COLUMN_DOES_NOT_EXITS: Key column 'type' doesn't exist in table`

Directus 8 screenshot:

![image](https://user-images.githubusercontent.com/156681/128043323-54d74459-8bd0-421a-a5f4-9635d5a57284.png)

moekify commented:
> Migration tool payload:
> 
> ```
>     url: '/relations',
>     method: 'post',
>     data: '{"meta":{"many_collection":"portfolio","many_field":"type","many_primary":"id","one_collection":"portfolio_types","one_field":null,"one_primary":"id","junction_field":null},"field":"type","collection":"portfolio","related_collection":"portfolio_types","schema":null}',
> ```
> 

@justrealmilk Where did you get this from? As I mentioned before, given your relations pictured in the screenshot. This is not the correct payload as is isn't referencing your junction collection :)

justrealmilk commented:
> > Migration tool payload:
> > ```
> >     url: '/relations',
> >     method: 'post',
> >     data: '{"meta":{"many_collection":"portfolio","many_field":"type","many_primary":"id","one_collection":"portfolio_types","one_field":null,"one_primary":"id","junction_field":null},"field":"type","collection":"portfolio","related_collection":"portfolio_types","schema":null}',
> > ```
> 
> @justrealmilk Where did you get this from? As I mentioned before, given your relations pictured in the screenshot. This is not the correct payload as is isn't referencing your junction collection :)

![image](https://user-images.githubusercontent.com/156681/128048644-a43b0b21-eb7c-44ed-a263-4e2638f3e3ca.png)

moekify commented:
Perhaps, it would make sense to have a look at your v8 relations output. It seems odd it tries to add this relation (at least to me from what I learned from debugging my own migration) 


I pushed a commit https://github.com/directus-community/migration-tool/pull/37/commits/22b0d2d105c33de6dd2a230d47daeed226c54329 that separates the getting from migrating. You can dumb your context in a pastebin or so and perhaps I can point out if there is something odd (to me)

justrealmilk commented:
> @justrealmilk Perhaps, it would make sense to have a look at your v8 relations output. It seems odd it tries to add this relation (at least to me from what I learned from debugging my own migration)
> 
> I pushed a commit [22b0d2d](https://github.com/directus-community/migration-tool/commit/22b0d2d105c33de6dd2a230d47daeed226c54329) that separates the getting from migrating. You can dumb your context in a pastebin or so and perhaps I can point out if there is something odd (to me)

I _think_ this is what you want

[relationsv8.txt](https://github.com/directus-community/migration-tool/files/6926371/relationsv8.txt)

moekify commented:
So, from what I can see your collections (schema) don't have the o2m fields? 

```
{
      "collection": "portfolio_types",
      "meta": {
        "note": null,
        "hidden": true,
        "singleton": false,
        "icon": null,
        "sort_field": null
      },
      "schema": {},
      "fields": [
        {
          "field": "id",
          "type": "integer",
          "meta": {
            "note": "",
            "interface": null,
            "readonly": false,
            "hidden": true,
            "width": null,
            "sort": 0
          },
          "schema": {
            "has_auto_increment": true,
            "default_value": null,
            "is_primary_key": true,
            "is_nullable": false,
            "max_length": "10",
            "numeric_precision": "10",
            "numeric_scale": null
          }
        },
        {
          "field": "name",
          "type": "string",
          "meta": {
            "note": "",
            "interface": "input",
            "readonly": false,
            "hidden": false,
            "width": "full",
            "sort": null
          },
          "schema": {
            "has_auto_increment": false,
            "default_value": null,
            "is_primary_key": false,
            "is_nullable": true,
            "max_length": "200",
            "numeric_precision": "200",
            "numeric_scale": null
          }
        }
      ]
    }
 ```
 
 Has no type field (which would be a o2m to the junction collection) 
 
 Did you make changes to your schema since you ran the migration tool?  Otherwise, there seems to be another issue, so that your collections aren't pulled correctly? 
 
 Would have nothing to do with this change :) 
 
 Perhaps, open another issue for that and maybe someone can take it up? Or you try yourself  :)  
 
 ---
 
 @rijkvanzanten I'd like to get this merged in, so I have it off my plate. Any other changes you see necessary? :)

justrealmilk commented:
> So, from what I can see your collections (schema) don't have the o2m fields?
> 
> ```
> {
>       "collection": "portfolio_types",
>       "meta": {
>         "note": null,
>         "hidden": true,
>         "singleton": false,
>         "icon": null,
>         "sort_field": null
>       },
>       "schema": {},
>       "fields": [
>         {
>           "field": "id",
>           "type": "integer",
>           "meta": {
>             "note": "",
>             "interface": null,
>             "readonly": false,
>             "hidden": true,
>             "width": null,
>             "sort": 0
>           },
>           "schema": {
>             "has_auto_increment": true,
>             "default_value": null,
>             "is_primary_key": true,
>             "is_nullable": false,
>             "max_length": "10",
>             "numeric_precision": "10",
>             "numeric_scale": null
>           }
>         },
>         {
>           "field": "name",
>           "type": "string",
>           "meta": {
>             "note": "",
>             "interface": "input",
>             "readonly": false,
>             "hidden": false,
>             "width": "full",
>             "sort": null
>           },
>           "schema": {
>             "has_auto_increment": false,
>             "default_value": null,
>             "is_primary_key": false,
>             "is_nullable": true,
>             "max_length": "200",
>             "numeric_precision": "200",
>             "numeric_scale": null
>           }
>         }
>       ]
>     }
> ```
> 
> Has no type field (which would be a o2m to the junction collection)
> 
> Did you make changes to your schema since you ran the migration tool? Otherwise, there seems to be another issue, so that your collections aren't pulled correctly?
> 
> Would have nothing to do with this change :)
> 
> Perhaps, open another issue for that and maybe someone can take it up? Or you try yourself :)
> 
> @rijkvanzanten I'd like to get this merged in, so I have it off my plate. Any other changes you see necessary? :)

That's one of the tables connected by the junction table - the m2m relation is between `portfolio` items (has a Directus field named `type`) and `portfolio_types`

moekify commented:
Sorry, I don't have the time to debug your v8 instance. if there is a m2m relationship, portfolio-types also should have a m2m field is what I was trying to say (but chose to say oddly I admit!)

Your v8 relations seem to have some "non-existent" relations. If you check the payload, you can see there is a relation that doesn't make sense between portfolio and portfolio_types + the relations with the junction collection. 

In any case, this is likely not an issue relating to the migration tool itself, but more to stale data in the v8 instance. FWIW, I had the exact same thing and had to clean up my data a bit (i.e. delete stale relations and junction collection records that were referencing non-existing items). 


```
{
      "collection_many": "portfolio",
      "collection_one": "portfolio_types",
      "field_many": "type",
      "field_one": null,
      "id": 15,
      "junction_field": null
    },
    {
      "collection_many": "portfolio_item_types",
      "collection_one": "portfolio",
      "field_many": "portfolio_id",
      "field_one": "type",
      "id": 16,
      "junction_field": "portfolio_types_id"
    },
    {
      "collection_many": "portfolio_item_types",
      "collection_one": "portfolio_types",
      "field_many": "portfolio_types_id",
      "field_one": "portfolio",
      "id": 17,
      "junction_field": "portfolio_id"
    },
   ``` 
  
First one cant work :) (Or at least I don't know how it would). 

Concluding, this is from my perspective unrelated to this PR and most likely related to your v8 data being a bit messy (Which I had to solve in my case too :) )

rijkvanzanten reviewed (approved):
> @rijkvanzanten I'd like to get this merged in, so I have it off my plate. Any other changes you see necessary? :)

Nope! LGTM

rijkvanzanten commented:
> [...] most likely related to your v8 data being a bit messy 

We can find ways to make the migration tool more robust for messy stuff in the v8 installation in subsequent PRs 👍🏻

ThaDaVos commented:
Am I right that this pull-request is only about saving the context? Not loading it back-in or even setting the `completedSteps`? Or am I using the wrong branch? I am on main, I see the files but they don't seem to be used and if I set the `completedSteps` manually I get errors about missing context properties, probably because the data isn't loaded back in

ThaDaVos commented:
Migration of the Relations seems to still fail if it's linked to `directus_files` as the id is now a uuid instead of a ID - does one really have to change this manually for all their collections? Or can this be picked up by the migration tool?

rijkvanzanten commented:
@ThaDaVos when the data is migrated, when a relationship to directus_files is encountered, it will replace the file ID with the new file UUID on the fly here:

https://github.com/directus-community/migration-tool/blob/db3d18dc4d5c18104271488f7fc364178a0396aa/tasks/data.js#L164-L166

ThaDaVos commented:
That does not change the collection, does it? As it's failing on a foreign key creation during `Relations Migration` as the custom collection still uses `int(11)` for it's column and not `char(36)`

I don't think this will be hard to change for myself inside the migration tool - maybe even add an option to convert all keys to UUIDS so one meets the new standard?

Currently working around it by running a small SQL script manually during the images transfer (as this takes some time) - to change all columns from `int(11)` to `char(36)` where it connects with `directus_files`

rijkvanzanten commented:
Ah I think it might be doing it for m2m fields to directus_files, but not for m2o 🤔 

https://github.com/directus-community/migration-tool/blob/main/tasks/schema.js#L180-L188
